From 6faebb8869308ec142bed9987e26fda1fdbcabd7 Mon Sep 17 00:00:00 2001
From: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
Date: Fri, 17 Feb 2023 04:46:52 -0800
Subject: [PATCH 4/4] ath12k: Optimize tx descriptor updates. Avoid
 FIELD_PREP

FIELD_PREP macros does a few sanity checks as part of field updates.
Use a simple method to update fields as it helps improve throughput

Signed-off-by: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>
Signed-off-by: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c | 32 ++++++++++----------------------
 1 file changed, 10 insertions(+), 22 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -207,18 +207,15 @@ int ath12k_dp_tx_direct(struct ath12k_li
 
 	ab->soc_stats.tx_enqueued[ring_id]++;
 
-	tcl_desc.buf_addr_info.info0 = FIELD_PREP(BUFFER_ADDR_INFO0_ADDR, skb_cb->paddr);
-	tcl_desc.buf_addr_info.info1 = FIELD_PREP(BUFFER_ADDR_INFO1_ADDR,
-			((uint64_t)skb_cb->paddr >> HAL_ADDR_MSB_REG_SHIFT));
-	tcl_desc.buf_addr_info.info1 |=
-		FIELD_PREP(BUFFER_ADDR_INFO1_SW_COOKIE, tx_desc->desc_id);
-	tcl_desc.info0 = FIELD_PREP(HAL_TCL_DATA_CMD_INFO0_DESC_TYPE,
-				    HAL_TCL_DESC_TYPE_BUFFER) | arvif->desc.info0;
+	tcl_desc.buf_addr_info.info0 = (u32)skb_cb->paddr;
+	tcl_desc.buf_addr_info.info1 = (tx_desc->desc_id << 12);
+	tcl_desc.info0 =  arvif->desc.info0;
 	tcl_desc.info1 =  arvif->desc.info1;
+	tcl_desc.info2 =  skb->len;
 
 	if (likely(skb->ip_summed == CHECKSUM_PARTIAL))
 		tcl_desc.info2 |= TX_IP_CHECKSUM;
-	tcl_desc.info2 |= FIELD_PREP(HAL_TCL_DATA_CMD_INFO2_DATA_LEN, skb->len);
+
 	tcl_desc.info3 = arvif->desc.info3;
 	tcl_desc.info4 =  arvif->desc.info4;
 	memcpy(hal_tcl_desc, &tcl_desc, sizeof(tcl_desc));
@@ -471,20 +468,12 @@ int ath12k_dp_tx(struct ath12k *ar, stru
 	arvif->link_stats.tx_desc_type[ti.type]++;
 	tcl_cmd = (struct hal_tcl_data_cmd *)hal_tcl_desc;
 
-	tcl_cmd->buf_addr_info.info0 =
-		FIELD_PREP(BUFFER_ADDR_INFO0_ADDR, ti.paddr);
-	tcl_cmd->buf_addr_info.info1 =
-		FIELD_PREP(BUFFER_ADDR_INFO1_ADDR,
-			   ((uint64_t)ti.paddr >> HAL_ADDR_MSB_REG_SHIFT));
-	tcl_cmd->buf_addr_info.info1 |=
-		FIELD_PREP(BUFFER_ADDR_INFO1_SW_COOKIE, ti.desc_id);
-
-	tcl_cmd->info0 = FIELD_PREP(HAL_TCL_DATA_CMD_INFO0_DESC_TYPE,
-				    ti.type) | arvif->desc.info0;
-	tcl_cmd->info1 = FIELD_PREP(HAL_TCL_DATA_CMD_INFO1_CMD_NUM,
-				    ti.meta_data_flags);
-	tcl_cmd->info2 = ti.flags0 |
-		FIELD_PREP(HAL_TCL_DATA_CMD_INFO2_DATA_LEN, ti.data_len);
+	tcl_cmd->buf_addr_info.info0 = (u32)ti.paddr;
+	/* TODO: Copy the upper 8 bits here */
+	tcl_cmd->buf_addr_info.info1 = (ti.desc_id << 12);
+	tcl_cmd->info0 = (ti.type << 1) | arvif->desc.info0;
+	tcl_cmd->info1 = ti.meta_data_flags << 16;
+	tcl_cmd->info2 = ti.flags0 | ti.data_len;
 
 	/* In tcl_cmd->info3, Bit 24 to 31 represents vdev_id
 	 * LSH 24 times to add updated vdev_id to info3
